File: sdk/go/license/license.go
// Package license provides TuskLang SDK License Validation Module
// Enterprise-grade license validation for Go SDK

package license

import (
	"crypto/hmac"
	"crypto/md5"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io/ioutil"
	"log"
	"net/http"
	"os"
	"path/filepath"
	"strconv"
	"strings"
	"sync"
	"time"

	"github.com/google/uuid"
)

// LicenseInfo represents comprehensive license information
type LicenseInfo struct {
	LicenseKey      string                 `json:"license_key"`
	SessionID       string                 `json:"session_id"`
	Validation      ValidationResult       `json:"validation"`
	Expiration      ExpirationResult       `json:"expiration"`
	CacheStatus     string                 `json:"cache_status"`
	ValidationCount int                    `json:"validation_count"`
	Warnings        int                    `json:"warnings"`
	CachedData      map[string]interface{} `json:"cached_data,omitempty"`
	CacheAge        int64                  `json:"cache_age,omitempty"`
}

// ValidationResult represents license validation result
type ValidationResult struct {
	Valid    bool   `json:"valid"`
	Error    string `json:"error,omitempty"`
	Checksum string `json:"checksum,omitempty"`
}

// ExpirationResult represents license expiration information
type ExpirationResult struct {
	Expired        bool   `json:"expired"`
	Error          string `json:"error,omitempty"`
	ExpirationDate string `json:"expiration_date,omitempty"`
	DaysRemaining  int64  `json:"days_remaining,omitempty"`
	DaysOverdue    int64  `json:"days_overdue,omitempty"`
	Warning        bool   `json:"warning,omitempty"`
}

// ValidationAttempt represents a license validation attempt
type ValidationAttempt struct {
	Timestamp int64  `json:"timestamp"`
	Success   bool   `json:"success"`
	Details   string `json:"details"`
	SessionID string `json:"session_id"`
}

// ExpirationWarning represents an expiration warning
type ExpirationWarning struct {
	Timestamp      int64 `json:"timestamp"`
	DaysRemaining  int64 `json:"days_remaining"`
}

// TuskLicense provides license validation functionality
type TuskLicense struct {
	licenseKey         string
	apiKey             string
	sessionID          string
	licenseCache       map[string]LicenseCacheEntry
	validationHistory  []ValidationAttempt
	expirationWarnings []ExpirationWarning
	mutex              sync.RWMutex
	httpClient         *http.Client
	cacheDir           string
	cacheFile          string
	offlineCache       *OfflineCacheData
	logger             *log.Logger
}

// OfflineCacheData represents offline cached license data
type OfflineCacheData struct {
	LicenseKeyHash string                 `json:"license_key_hash"`
	LicenseData    map[string]interface{} `json:"license_data"`
	Timestamp      int64                  `json:"timestamp"`
	Expiration     ExpirationResult       `json:"expiration"`
}

// LicenseCacheEntry represents cached license data
type LicenseCacheEntry struct {
	Data      map[string]interface{} `json:"data"`
	Timestamp int64                  `json:"timestamp"`
	Expires   int64                  `json:"expires"`
}

// New creates a new TuskLicense instance
func New(licenseKey, apiKey string) *TuskLicense {
	return NewWithCacheDir(licenseKey, apiKey, "")
}

// NewWithCacheDir creates a new TuskLicense instance with custom cache directory
func NewWithCacheDir(licenseKey, apiKey, cacheDir string) *TuskLicense {
	// Set up cache directory
	if cacheDir == "" {
		homeDir, _ := os.UserHomeDir()
		cacheDir = filepath.Join(homeDir, ".tusk", "license_cache")
	}

	// Create cache directory if it doesn't exist
	os.MkdirAll(cacheDir, 0755)

	// Generate cache file name based on license key hash
	keyHash := md5.Sum([]byte(licenseKey))
	cacheFile := filepath.Join(cacheDir, fmt.Sprintf("%x.cache", keyHash))

	tl := &TuskLicense{
		licenseKey:         licenseKey,
		apiKey:             apiKey,
		sessionID:          uuid.New().String(),
		licenseCache:       make(map[string]LicenseCacheEntry),
		validationHistory:  make([]ValidationAttempt, 0),
		expirationWarnings: make([]ExpirationWarning, 0),
		httpClient:         &http.Client{Timeout: 10 * time.Second},
		cacheDir:           cacheDir,
		cacheFile:          cacheFile,
		logger:             log.New(os.Stderr, "[TuskLicense] ", log.LstdFlags),
	}

	// Load offline cache if exists
	tl.loadOfflineCache()

	return tl
}

// ValidateLicenseKey validates license key format and checksum
func (tl *TuskLicense) ValidateLicenseKey() ValidationResult {
	if len(tl.licenseKey) < 32 {
		return ValidationResult{
			Valid: false,
			Error: "Invalid license key format",
		}
	}

	if !strings.HasPrefix(tl.licenseKey, "TUSK-") {
		return ValidationResult{
			Valid: false,
			Error: "Invalid license key prefix",
		}
	}

	hash := sha256.Sum256([]byte(tl.licenseKey))
	checksum := hex.EncodeToString(hash[:])

	if !strings.HasPrefix(checksum, "tusk") {
		return ValidationResult{
			Valid: false,
			Error: "Invalid license key checksum",
		}
	}

	return ValidationResult{
		Valid:    true,
		Checksum: checksum,
	}
}

// VerifyLicenseServer verifies license with remote server
func (tl *TuskLicense) VerifyLicenseServer(serverURL string) (map[string]interface{}, error) {
	if serverURL == "" {
		serverURL = "https://api.tusklang.org/v1/license"
	}

	timestamp := time.Now().Unix()
	data := map[string]interface{}{
		"license_key": tl.licenseKey,
		"session_id":  tl.sessionID,
		"timestamp":   timestamp,
	}

	// Generate signature
	jsonData, err := json.Marshal(data)
	if err != nil {
		return nil, fmt.Errorf("failed to marshal data: %w", err)
	}

	h := hmac.New(sha256.New, []byte(tl.apiKey))
	h.Write(jsonData)
	signature := hex.EncodeToString(h.Sum(nil))
	data["signature"] = signature

	// Make HTTP request
	jsonPayload, err := json.Marshal(data)
	if err != nil {
		return nil, fmt.Errorf("failed to marshal payload: %w", err)
	}

	req, err := http.NewRequest("POST", serverURL, strings.NewReader(string(jsonPayload)))
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}

	req.Header.Set("Authorization", "Bearer "+tl.apiKey)
	req.Header.Set("Content-Type", "application/json")

	resp, err := tl.httpClient.Do(req)
	if err != nil {
		tl.logger.Printf("Network error during license validation: %v\n", err)
		return tl.fallbackToOfflineCache(fmt.Sprintf("network error: %v", err))
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		tl.logger.Printf("Server returned error: %d\n", resp.StatusCode)
		return tl.fallbackToOfflineCache(fmt.Sprintf("server error: %d", resp.StatusCode))
	}

	var result map[string]interface{}
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, fmt.Errorf("failed to decode response: %w", err)
	}

	// Cache the result
	tl.mutex.Lock()
	tl.licenseCache[tl.licenseKey] = LicenseCacheEntry{
		Data:      result,
		Timestamp: timestamp,
		Expires:   timestamp + 3600, // 1 hour cache
	}
	tl.mutex.Unlock()

	// Save to offline cache
	tl.saveOfflineCache(result)

	return result, nil
}

// CheckLicenseExpiration checks if license is expired or expiring soon
func (tl *TuskLicense) CheckLicenseExpiration() ExpirationResult {
	parts := strings.Split(tl.licenseKey, "-")
	if len(parts) < 4 {
		return ExpirationResult{
			Expired: true,
			Error:   "Invalid license key format",
		}
	}

	expirationStr := parts[len(parts)-1]
	expirationTimestamp, err := strconv.ParseInt(expirationStr, 16, 64)
	if err != nil {
		return ExpirationResult{
			Expired: true,
			Error:   "Invalid expiration timestamp",
		}
	}

	expirationDate := time.Unix(expirationTimestamp, 0)
	currentTime := time.Now()

	if expirationDate.Before(currentTime) {
		daysOverdue := int64(currentTime.Sub(expirationDate).Hours() / 24)
		return ExpirationResult{
			Expired:        true,
			ExpirationDate: expirationDate.Format(time.RFC3339),
			DaysOverdue:    daysOverdue,
		}
	}

	daysRemaining := int64(expirationDate.Sub(currentTime).Hours() / 24)

	if daysRemaining <= 30 {
		tl.mutex.Lock()
		tl.expirationWarnings = append(tl.expirationWarnings, ExpirationWarning{
			Timestamp:     currentTime.Unix(),
			DaysRemaining: daysRemaining,
		})
		tl.mutex.Unlock()
	}

	return ExpirationResult{
		Expired:        false,
		ExpirationDate: expirationDate.Format(time.RFC3339),
		DaysRemaining:  daysRemaining,
		Warning:        daysRemaining <= 30,
	}
}

// ValidateLicensePermissions validates if license allows specific feature
func (tl *TuskLicense) ValidateLicensePermissions(feature string) (bool, error) {
	tl.mutex.RLock()
	if cacheEntry, exists := tl.licenseCache[tl.licenseKey]; exists {
		if time.Now().Unix() < cacheEntry.Expires {
			if features, ok := cacheEntry.Data["features"].([]interface{}); ok {
				for _, f := range features {
					if fStr, ok := f.(string); ok && fStr == feature {
						tl.mutex.RUnlock()
						return true, nil
					}
				}
				tl.mutex.RUnlock()
				return false, fmt.Errorf("feature not licensed")
			}
		}
	}
	tl.mutex.RUnlock()

	// Fallback validation
	switch feature {
	case "basic", "core", "standard":
		return true, nil
	case "premium", "enterprise":
		upperKey := strings.ToUpper(tl.licenseKey)
		if strings.Contains(upperKey, "PREMIUM") || strings.Contains(upperKey, "ENTERPRISE") {
			return true, nil
		}
		return false, fmt.Errorf("premium license required")
	default:
		return false, fmt.Errorf("unknown feature")
	}
}

// GetLicenseInfo returns comprehensive license information
func (tl *TuskLicense) GetLicenseInfo() LicenseInfo {
	validationResult := tl.ValidateLicenseKey()
	expirationResult := tl.CheckLicenseExpiration()

	tl.mutex.RLock()
	cacheStatus := "not_cached"
	var cachedData map[string]interface{}
	var cacheAge int64

	if cacheEntry, exists := tl.licenseCache[tl.licenseKey]; exists {
		cacheStatus = "cached"
		cachedData = cacheEntry.Data
		cacheAge = time.Now().Unix() - cacheEntry.Timestamp
	}

	info := LicenseInfo{
		LicenseKey:      fmt.Sprintf("%s...%s", tl.licenseKey[:8], tl.licenseKey[len(tl.licenseKey)-4:]),
		SessionID:       tl.sessionID,
		Validation:      validationResult,
		Expiration:      expirationResult,
		CacheStatus:     cacheStatus,
		ValidationCount: len(tl.validationHistory),
		Warnings:        len(tl.expirationWarnings),
		CachedData:      cachedData,
		CacheAge:        cacheAge,
	}
	tl.mutex.RUnlock()

	return info
}

// LogValidationAttempt logs a license validation attempt
func (tl *TuskLicense) LogValidationAttempt(success bool, details string) {
	tl.mutex.Lock()
	tl.validationHistory = append(tl.validationHistory, ValidationAttempt{
		Timestamp: time.Now().Unix(),
		Success:   success,
		Details:   details,
		SessionID: tl.sessionID,
	})
	tl.mutex.Unlock()
}

// GetValidationHistory returns validation history
func (tl *TuskLicense) GetValidationHistory() []ValidationAttempt {
	tl.mutex.RLock()
	defer tl.mutex.RUnlock()
	
	history := make([]ValidationAttempt, len(tl.validationHistory))
	copy(history, tl.validationHistory)
	return history
}

// ClearValidationHistory clears validation history
func (tl *TuskLicense) ClearValidationHistory() {
	tl.mutex.Lock()
	tl.validationHistory = tl.validationHistory[:0]
	tl.mutex.Unlock()
}

// Global license instance
var (
	licenseInstance *TuskLicense
	instanceMutex   sync.RWMutex
)

// InitializeLicense creates the global license instance
func InitializeLicense(licenseKey, apiKey string) *TuskLicense {
	instanceMutex.Lock()
	defer instanceMutex.Unlock()

	licenseInstance = New(licenseKey, apiKey)
	return licenseInstance
}

// GetLicense returns the global license instance
func GetLicense() *TuskLicense {
	instanceMutex.RLock()
	defer instanceMutex.RUnlock()

	if licenseInstance == nil {
		panic("License not initialized. Call InitializeLicense() first.")
	}
	return licenseInstance
}

// loadOfflineCache loads offline license cache from disk
func (tl *TuskLicense) loadOfflineCache() {
	data, err := ioutil.ReadFile(tl.cacheFile)
	if err != nil {
		// Cache file doesn't exist or can't be read
		tl.offlineCache = nil
		return
	}

	var cached OfflineCacheData
	if err := json.Unmarshal(data, &cached); err != nil {
		tl.logger.Printf("Failed to parse offline cache: %v\n", err)
		tl.offlineCache = nil
		return
	}

	// Verify the cache is for the correct license key
	hash := sha256.Sum256([]byte(tl.licenseKey))
	keyHash := hex.EncodeToString(hash[:])
	if cached.LicenseKeyHash == keyHash {
		tl.offlineCache = &cached
		tl.logger.Println("Loaded offline license cache")
	} else {
		tl.logger.Println("Offline cache key mismatch")
		tl.offlineCache = nil
	}
}

// saveOfflineCache saves license data to offline cache
func (tl *TuskLicense) saveOfflineCache(licenseData map[string]interface{}) {
	hash := sha256.Sum256([]byte(tl.licenseKey))
	keyHash := hex.EncodeToString(hash[:])

	cacheData := OfflineCacheData{
		LicenseKeyHash: keyHash,
		LicenseData:    licenseData,
		Timestamp:      time.Now().Unix(),
		Expiration:     tl.CheckLicenseExpiration(),
	}

	data, err := json.MarshalIndent(cacheData, "", "  ")
	if err != nil {
		tl.logger.Printf("Failed to marshal cache data: %v\n", err)
		return
	}

	if err := ioutil.WriteFile(tl.cacheFile, data, 0600); err != nil {
		tl.logger.Printf("Failed to save offline cache: %v\n", err)
		return
	}

	tl.offlineCache = &cacheData
	tl.logger.Println("Saved license data to offline cache")
}

// fallbackToOfflineCache fallback to offline cache when server is unreachable
func (tl *TuskLicense) fallbackToOfflineCache(errorMsg string) (map[string]interface{}, error) {
	if tl.offlineCache != nil && tl.offlineCache.LicenseData != nil {
		cacheAge := time.Now().Unix() - tl.offlineCache.Timestamp
		cacheAgeDays := float64(cacheAge) / 86400.0

		// Check if cached license is not expired
		if !tl.offlineCache.Expiration.Expired {
			tl.logger.Printf("Using offline license cache (age: %.1f days)\n", cacheAgeDays)
			result := make(map[string]interface{})
			for k, v := range tl.offlineCache.LicenseData {
				result[k] = v
			}
			result["offline_mode"] = true
			result["cache_age_days"] = cacheAgeDays
			result["warning"] = fmt.Sprintf("Operating in offline mode due to: %s", errorMsg)
			return result, nil
		}
		return nil, fmt.Errorf("license expired and server unreachable: %s", errorMsg)
	}
	return nil, fmt.Errorf("no offline cache available: %s", errorMsg)
} 

File: sdk/javascript/tusk/license.js
/**
 * TuskLang SDK License Validation Module
 * Enterprise-grade license validation for JavaScript SDK
 */

const crypto = require('crypto');
const https = require('https');
const http = require('http');
const fs = require('fs').promises;
const path = require('path');
const os = require('os');

class TuskLicense {
    constructor(licenseKey, apiKey, cacheDir = null) {
        this.licenseKey = licenseKey;
        this.apiKey = apiKey;
        this.sessionId = this.generateUUID();
        this.licenseCache = new Map();
        this.validationHistory = [];
        this.expirationWarnings = [];
        
        // Set up offline cache directory
        if (cacheDir) {
            this.cacheDir = cacheDir;
        } else {
            this.cacheDir = path.join(os.homedir(), '.tusk', 'license_cache');
        }
        this.cacheFile = path.join(this.cacheDir, `${crypto.createHash('md5').update(licenseKey).digest('hex')}.cache`);
        
        // Load offline cache if exists
        this.offlineCache = null;
        this._loadOfflineCache();
    }

    generateUUID() {
        return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, function(c) {
            const r = Math.random() * 16 | 0;
            const v = c == 'x' ? r : (r & 0x3 | 0x8);
            return v.toString(16);
        });
    }

    validateLicenseKey() {
        try {
            if (!this.licenseKey || this.licenseKey.length < 32) {
                return { valid: false, error: 'Invalid license key format' };
            }

            if (!this.licenseKey.startsWith('TUSK-')) {
                return { valid: false, error: 'Invalid license key prefix' };
            }

            const checksum = crypto.createHash('sha256').update(this.licenseKey).digest('hex');
            if (!checksum.startsWith('tusk')) {
                return { valid: false, error: 'Invalid license key checksum' };
            }

            return { valid: true, checksum };
        } catch (error) {
            return { valid: false, error: error.message };
        }
    }

    async verifyLicenseServer(serverUrl = 'https://api.tusklang.org/v1/license') {
        return new Promise((resolve) => {
            try {
                const data = {
                    license_key: this.licenseKey,
                    session_id: this.sessionId,
                    timestamp: Math.floor(Date.now() / 1000)
                };

                const signature = crypto
                    .createHmac('sha256', this.apiKey)
                    .update(JSON.stringify(data))
                    .digest('hex');
                data.signature = signature;

                const postData = JSON.stringify(data);
                const url = new URL(serverUrl);
                const options = {
                    hostname: url.hostname,
                    port: url.port || (url.protocol === 'https:' ? 443 : 80),
                    path: url.pathname,
                    method: 'POST',
                    headers: {
                        'Authorization': `Bearer ${this.apiKey}`,
                        'Content-Type': 'application/json',
                        'Content-Length': Buffer.byteLength(postData)
                    }
                };

                const client = url.protocol === 'https:' ? https : http;
                const req = client.request(options, (res) => {
                    let responseData = '';
                    res.on('data', (chunk) => {
                        responseData += chunk;
                    });
                    res.on('end', async () => {
                        if (res.statusCode === 200) {
                            try {
                                const result = JSON.parse(responseData);
                                this.licenseCache.set(this.licenseKey, {
                                    data: result,
                                    timestamp: Date.now(),
                                    expires: Date.now() + 3600000 // 1 hour
                                });
                                // Save to offline cache
                                await this._saveOfflineCache(result);
                                resolve(result);
                            } catch (error) {
                                console.warn('Failed to parse server response:', error);
                                resolve(await this._fallbackToOfflineCache('Invalid response format'));
                            }
                        } else {
                            console.warn(`Server returned error: ${res.statusCode}`);
                            resolve(await this._fallbackToOfflineCache(`Server error: ${res.statusCode}`));
                        }
                    });
                });

                req.on('error', async (error) => {
                    console.warn('Network error during license validation:', error.message);
                    resolve(await this._fallbackToOfflineCache(`Network error: ${error.message}`));
                });

                req.setTimeout(10000, async () => {
                    req.destroy();
                    console.warn('License validation request timeout');
                    resolve(await this._fallbackToOfflineCache('Request timeout'));
                });

                req.write(postData);
                req.end();
            } catch (error) {
                console.error('Unexpected error during license validation:', error);
                resolve(await this._fallbackToOfflineCache(error.message));
            }
        });
    }
    
    async _loadOfflineCache() {
        try {
            await fs.mkdir(this.cacheDir, { recursive: true });
            const cacheData = await fs.readFile(this.cacheFile, 'utf8');
            const cached = JSON.parse(cacheData);
            
            // Verify the cache is for the correct license key
            const keyHash = crypto.createHash('sha256').update(this.licenseKey).digest('hex');
            if (cached.licenseKeyHash === keyHash) {
                this.offlineCache = cached;
                console.info('Loaded offline license cache');
            } else {
                console.warn('Offline cache key mismatch');
                this.offlineCache = null;
            }
        } catch (error) {
            // Cache file doesn't exist or is invalid
            this.offlineCache = null;
        }
    }
    
    async _saveOfflineCache(licenseData) {
        try {
            const cacheData = {
                licenseKeyHash: crypto.createHash('sha256').update(this.licenseKey).digest('hex'),
                licenseData: licenseData,
                timestamp: Date.now(),
                expiration: this.checkLicenseExpiration()
            };
            
            await fs.mkdir(this.cacheDir, { recursive: true });
            await fs.writeFile(this.cacheFile, JSON.stringify(cacheData, null, 2));
            this.offlineCache = cacheData;
            console.info('Saved license data to offline cache');
        } catch (error) {
            console.error('Failed to save offline cache:', error);
        }
    }
    
    async _fallbackToOfflineCache(errorMsg) {
        if (this.offlineCache && this.offlineCache.licenseData) {
            const cacheAge = Date.now() - this.offlineCache.timestamp;
            const cacheAgeDays = cacheAge / 86400000;
            
            // Check if cached license is not expired
            const expiration = this.offlineCache.expiration || {};
            if (!expiration.expired) {
                console.warn(`Using offline license cache (age: ${cacheAgeDays.toFixed(1)} days)`);
                return {
                    ...this.offlineCache.licenseData,
                    offline_mode: true,
                    cache_age_days: cacheAgeDays,
                    warning: `Operating in offline mode due to: ${errorMsg}`
                };
            } else {
                return {
                    valid: false,
                    error: `License expired and server unreachable: ${errorMsg}`,
                    offline_cache_expired: true
                };
            }
        } else {
            return {
                valid: false,
                error: `No offline cache available: ${errorMsg}`,
                offline_cache_missing: true
            };
        }
    }

    checkLicenseExpiration() {
        try {
            const parts = this.licenseKey.split('-');
            if (parts.length < 4) {
                return { expired: true, error: 'Invalid license key format' };
            }

            const expirationStr = parts[parts.length - 1];
            const expirationTimestamp = parseInt(expirationStr, 16);
            const expirationDate = new Date(expirationTimestamp * 1000);
            const currentDate = new Date();

            if (expirationDate < currentDate) {
                return {
                    expired: true,
                    expiration_date: expirationDate.toISOString(),
                    days_overdue: Math.floor((currentDate - expirationDate) / (1000 * 60 * 60 * 24))
                };
            }

            const daysUntilExpiration = Math.floor((expirationDate - currentDate) / (1000 * 60 * 60 * 24));

            if (daysUntilExpiration <= 30) {
                this.expirationWarnings.push({
                    timestamp: Date.now(),
                    days_remaining: daysUntilExpiration
                });
            }

            return {
                expired: false,
                expiration_date: expirationDate.toISOString(),
                days_remaining: daysUntilExpiration,
                warning: daysUntilExpiration <= 30
            };
        } catch (error) {
            return { expired: true, error: error.message };
        }
    }

    validateLicensePermissions(feature) {
        try {
            if (this.licenseCache.has(this.licenseKey)) {
                const cacheData = this.licenseCache.get(this.licenseKey);
                if (Date.now() < cacheData.expires) {
                    const licenseData = cacheData.data;
                    if (licenseData.features) {
                        const allowedFeatures = licenseData.features;
                        if (allowedFeatures.includes(feature)) {
                            return { allowed: true, feature };
                        } else {
                            return { allowed: false, feature, error: 'Feature not licensed' };
                        }
                    }
                }
            }

            if (['basic', 'core', 'standard'].includes(feature)) {
                return { allowed: true, feature };
            } else if (['premium', 'enterprise'].includes(feature)) {
                if (this.licenseKey.toUpperCase().includes('PREMIUM') || 
                    this.licenseKey.toUpperCase().includes('ENTERPRISE')) {
                    return { allowed: true, feature };
                } else {
                    return { allowed: false, feature, error: 'Premium license required' };
                }
            } else {
                return { allowed: false, feature, error: 'Unknown feature' };
            }
        } catch (error) {
            return { allowed: false, feature, error: error.message };
        }
    }

    getLicenseInfo() {
        try {
            const validationResult = this.validateLicenseKey();
            const expirationResult = this.checkLicenseExpiration();

            const info = {
                license_key: this.licenseKey.substring(0, 8) + '...' + this.licenseKey.substring(this.licenseKey.length - 4),
                session_id: this.sessionId,
                validation: validationResult,
                expiration: expirationResult,
                cache_status: this.licenseCache.has(this.licenseKey) ? 'cached' : 'not_cached',
                validation_count: this.validationHistory.length,
                warnings: this.expirationWarnings.length
            };

            if (this.licenseCache.has(this.licenseKey)) {
                const cacheData = this.licenseCache.get(this.licenseKey);
                info.cached_data = cacheData.data;
                info.cache_age = Date.now() - cacheData.timestamp;
            }

            return info;
        } catch (error) {
            return { error: error.message };
        }
    }

    refreshLicenseCache() {
        return this.verifyLicenseServer();
    }

    logValidationAttempt(success, details = '') {
        this.validationHistory.push({
            timestamp: Date.now(),
            success,
            details,
            session_id: this.sessionId
        });
    }

    getValidationHistory() {
        return this.validationHistory;
    }

    clearValidationHistory() {
        this.validationHistory = [];
    }
}

// Global license instance
let licenseInstance = null;

function initializeLicense(licenseKey, apiKey, cacheDir = null) {
    licenseInstance = new TuskLicense(licenseKey, apiKey, cacheDir);
    return licenseInstance;
}

function getLicense() {
    if (!licenseInstance) {
        throw new Error('License not initialized. Call initializeLicense() first.');
    }
    return licenseInstance;
}

module.exports = {
    TuskLicense,
    initializeLicense,
    getLicense
}; 